substratevm/mx.substratevm/testhello.py

```diff
@@ -132,20 +132,33 @@ def test():
     wildcard_pattern = '.*'
     # disable prompting to continue output
     execute("set pagination off")
+
+    # set a break point at java.lang.System::identityHashCode
+    # expect "Breakpoint 1 at 0x[0-9a-f]+: file com/oracle/svm/core/jdk/JavaLangSubstitutions.java, line [0-9]+"
+    exec_string = execute("break java.lang.System::identityHashCode")
```

Comment:
This is too fragile.

Instead please add method to `hello.Hello` and a corresponding method-substitution for it that is fully under your control and set a breakpoint to that. This way we don't have to adjust this verification code every time we change something in `JavaLangSubstitutions.java`

Reply:
@olpaw could you please point me to documentation or an example on how to register my substitution methods?

I have been looking at `/compiler/src/org.graalvm.compiler.replacements.test/src/org/graalvm/compiler/replacements/test/SubstitutionsTest.java` but tracing the registration of the method substitution seems more complex than what I would expect, do I need to do the same or is there any simpler way?

Thanks

Reply:
How about:
```patch
diff --git a/substratevm/mx.substratevm/suite.py b/substratevm/mx.substratevm/suite.py
index a06037a91ff..75063b4edea 100644
--- a/substratevm/mx.substratevm/suite.py
+++ b/substratevm/mx.substratevm/suite.py
@@ -560,7 +560,7 @@ suite = {
             "sourceDirs": ["src"],
             "dependencies": [
                 "mx:JUNIT_TOOL",
-                "sdk:GRAAL_SDK",
+                "SVM",
             ],
             "checkstyle": "com.oracle.svm.core",
             "workingSets": "SVM",
@@ -1129,7 +1129,7 @@ suite = {
           ],
           "distDependencies": [
             "mx:JUNIT_TOOL",
-            "sdk:GRAAL_SDK",
+            "SVM",
           ],
           "testDistribution" : True,
         },
diff --git a/substratevm/src/com.oracle.svm.test/src/hello/Hello.java b/substratevm/src/com.oracle.svm.test/src/hello/Hello.java
index 0482024ce1a..b23c858a925 100644
--- a/substratevm/src/com.oracle.svm.test/src/hello/Hello.java
+++ b/substratevm/src/com.oracle.svm.test/src/hello/Hello.java
@@ -31,15 +31,20 @@ package hello;
 public class Hello {
     public abstract static class Greeter {
         static Greeter greeter(String[] args) {
-            if (args.length == 0) {
+            int greeterID = greeterID(args);
+            if (greeterID == 0) {
                 return new DefaultGreeter();
-            } else if (args.length == 1) {
+            } else if (greeterID == 1) {
                 return new NamedGreeter(args[0]);
             } else {
                 throw new RuntimeException("zero or one args expected");
             }
         }
 
+        public static int greeterID(String[] args) {
+            throw new UnsupportedOperationException();
+        }
+
         public abstract void greet();
     }
 
diff --git a/substratevm/src/com.oracle.svm.test/src/hello/Target_hello_Hello_Greeter.java b/substratevm/src/com.oracle.svm.test/src/hello/Target_hello_Hello_Greeter.java
new file mode 100644
index 00000000000..c88d2b243dc
--- /dev/null
+++ b/substratevm/src/com.oracle.svm.test/src/hello/Target_hello_Hello_Greeter.java
@@ -0,0 +1,12 @@
+package hello;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(value = Hello.Greeter.class)
+final class Target_hello_Hello_Greeter {
+    @Substitute
+    static int greeterID(String[] args) {
+        return args.length;
+    }
+}
```

Reply:
That's great. Using the internal API makes it so easy :). I was using `org.graalvm.compiler.api.replacements.ClassSubstitution` and `org.graalvm.compiler.api.replacements.MethodSubstitution` instead :/

Thanks @olpaw! 

Note: The new test unveiled an issue which I am still working on.